dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -178,7 +178,14 @@ def sign_up_google_oauth2
     # our tracking data is usually populated, so do it here
     SignUpTracking.begin_sign_up_tracking(session, split_test: true)
 
-    user = User.from_omniauth auth_hash, auth_params, session
+    user =
+      if SignUpTracking.new_sign_up_experience?(session)
+        User.new.tap do |u|
+          User.initialize_new_oauth_user(u, auth_hash, auth_params)
```

Comment:
i think using `initialize_new_oauth_user` instead of `from_omniauth` will result in the same tokens-not-being-updated issue we were seeing this morning ([link](https://github.com/code-dot-org/code-dot-org/pull/24975) for future reference).

this is because `from_omniauth` [sets oauth tokens on the user](https://github.com/code-dot-org/code-dot-org/blob/cd61c8f12a3de0d3ff3562691dbb4e39b0dab423/dashboard/app/models/user.rb#L721), but [`initialize_new_oauth_user`](https://github.com/code-dot-org/code-dot-org/blob/cd61c8f12a3de0d3ff3562691dbb4e39b0dab423/dashboard/app/models/user.rb#L725-L765) does not

Reply:
:face_with_head_bandage: Yep, that's totally true!  Thanks for catching that here.  I'll fix up this PR and ask for another review.

Reply:
Ugh, as I get into this I realize it's a more painful bug than I thought.  We can't just reuse the current approach to setting oauth tokens on the created user because sometimes - and always under the new finish_sign_up flow - we haven't yet _created_ the user.

This means I'll have to go ahead and do the work now to push the oauth tokens into the session (and shared_cache) directly instead of setting them on the unsaved user and then pushing all of the user attributes into the session.  We'd have to do this eventually anyway, because the current approach isn't multi-auth friendly!